File: php-version/godaddydns.php
<?php
date_default_timezone_set("GMT");

$dir = dirname(dirname(__FILE__));
#根域名列表文件，如果自己的根域名不存在该文件中，可自行添加
$domainfile = $dir . DIRECTORY_SEPARATOR . "domain.ini";

# 第一个参数是 action，代表 (add/clean) 
# 第二个参数是域名 
# 第三个参数是主机名（第三个参数+第二个参数组合起来就是要添加的 TXT 记录）
# 第四个参数是 TXT 记录值
# 第五个参数是 APPKEY
# 第六个参数是 APPTOKEN

echo "域名 API 调用开始\n";
print_r($argv);
if (count($argv) < 7) {
    echo "参数有误\n";
    exit;
}
echo $argv[1]."-".$argv[2]."-".$argv[3]."-".$argv[4]."-".$argv[5]."-".$argv[6]."\n";

$domainarray = GodaddyDns::getDomain($argv[2]);
$selfdomain  = ($domainarray[0] == "") ? $argv[3] : $argv[3].".".$domainarray[0];

/*

  $obj = new GodaddyDns($argv[5], $argv[6], $domainarray[1]);
  $data = $obj->getDomains();
  $data_obj = json_decode($data['result']);
  $code = $data['httpCode'];
  test :php  godaddydns.php add yudadan.com  v k
 */

$obj = new GodaddyDns($argv[5], $argv[6], $domainarray[1]);

switch ($argv[1]) {
    case "clean":
        //api 不包含该操作
        break;

    case "add":
        //$data     = $obj->GetDNSRecord($domainarray[1], $selfdomain);
        //$data_obj = json_decode($data['result']);
        //$count    = count($data_obj);
        //if ($count > 0) {

        //    $data = $obj->UpdateDNSRecord($domainarray[1], $selfdomain, $argv[4]);
        //} else {
            $data = $obj->CreateDNSRecord($domainarray[1], $selfdomain, $argv[4]);
        //}
        if ($data["httpCode"] != 200) {
            $message = json_decode($data["result"], true);
            echo "域名处理失败-".$message["message"];
            exit;
        }
        break;
}

echo "域名 API 调用结束\n";

//    $r = $obj->UpdateDNSRecord($domain, $selfdomain, $argv[3], $type);

class GodaddyDns
{
    private $accessKeyId  = null;
    private $accessSecrec = null;
    private $DomainName   = null;
    private $Host         = "";
    private $Path         = "";

    public function __construct($accessKeyId, $accessSecrec, $domain = "")
    {
        $this->accessKeyId  = $accessKeyId;
        $this->accessSecrec = $accessSecrec;
        $this->DomainName   = $domain;
    }
    /*
      根据域名返回主机名和二级域名
     */

    public static function getDomain($domain)
    {

        //常见根域名 【https://en.wikipedia.org/wiki/List_of_Internet_top-level_domains】
        // 【http://www.seobythesea.com/2006/01/googles-most-popular-and-least-popular-top-level-domains/】

	global $domainfile;
	$tmp = file($domainfile);
	$arr = array();
	foreach ($tmp as $k=>$v) {
		$v = trim($v);
		if ($v!="")
			$arr[]= "." . $v;
	}

        //二级域名
        $seconddomain = "";
        //子域名
        $selfdomain   = "";
        //根域名
        $rootdomain   = "";
        foreach ($arr as $k => $v) {
            $pos = stripos($domain, $v);
            if ($pos) {
                $rootdomain   = substr($domain, $pos);
                $s            = explode(".", substr($domain, 0, $pos));
                $seconddomain = $s[count($s) - 1].$rootdomain;
                for ($i = 0; $i < count($s) - 1; $i++)
                    $selfdomain .= $s[$i] . ".";
		$selfdomain = substr($selfdomain,0,strlen($selfdomain)-1);
		break;
            }
        }
        //echo $seconddomain ;exit;
        if ($rootdomain == "") {
            $seconddomain = $domain;
            $selfdomain   = "";
        }
        return array($selfdomain, $seconddomain);
    }

    public function error($code, $str)
    {
        echo "操作错误:".$code.":".$str;
        exit;
    }

    private function curl($url, $header = '', $data = '', $method = 'get')
    {
        $ch       = curl_init();
        curl_setopt($ch, CURLOPT_URL, $url);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
        curl_setopt($ch, CURLOPT_CUSTOMREQUEST, $method); //设置请求方式
        curl_setopt($ch, CURLOPT_HTTPHEADER, $header);
        curl_setopt($ch, CURLOPT_POSTFIELDS, $data); //设置提交的字符串
        $result   = curl_exec($ch);
        $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);
        return array(
            'result' => $result,
            'httpCode' => $httpCode
        );
    }

    private function out($msg)
    {
        return json_decode($msg, true);
    }

    public function getDomains()
    {

        $url    = "https://api.godaddy.com/v1/domains";
        $header = ['accept: application/json', 'authorization:sso-key '.$this->accessKeyId.':'.$this->accessSecrec];
        return $this->curl($url, $header);
    }

    public function delRecords($domain)
    {

        $url    = "https://api.godaddy.com/v1/domains/$domain";
        $header = ['accept: application/json', 'Content-Type: application/json',
            'authorization:sso-key '.$this->accessKeyId.':'.$this->accessSecrec];

        return $this->curl($url, $header, '', 'delete');
    }

    public function GetDNSRecord($domain, $record, $recordType = 'TXT')
    {
        $url    = "https://api.godaddy.com/v1/domains/$domain/records/$recordType/$record";
        $header = ['accept: application/json', 'authorization:sso-key '.$this->accessKeyId.':'.$this->accessSecrec];
        return $this->curl($url, $header);
    }

    public function UpdateDNSRecord($domain, $name, $value, $recordType = 'TXT')
    {
        $url    = "https://api.godaddy.com/v1/domains/$domain/records/$recordType/$name";
        $header = ['accept: application/json', 'Content-Type: application/json',
            'authorization:sso-key '.$this->accessKeyId.':'.$this->accessSecrec];
        $data   = array(
            array(
                'data' => $value,
                'name' => $name,
                'ttl' => 3600,
                'type' => $recordType)
        );
        return $this->curl($url, $header, json_encode($data), 'put');
    }

    public function CreateDNSRecord($domain, $name, $value, $recordType = 'TXT')
    {
        $url    = "https://api.godaddy.com/v1/domains/$domain/records";
        $header = ['accept: application/json', 'Content-Type: application/json',
            'authorization:sso-key '.$this->accessKeyId.':'.$this->accessSecrec];
        $data   = array(
            array(
                'data' => $value,
                'name' => $name,
                'ttl' => 3600,
                'type' => $recordType)
        );
        return $this->curl($url, $header, json_encode($data), 'PATCH');
    }
}

File: python-version/godaddydns.py
# coding:utf-8

import json
import sys
import os 

class GodaddyDns:
    def __init__(self, access_key_id, access_key_secret, domain_name):
        self.access_key_id = access_key_id
        self.access_key_secret = access_key_secret
        self.domain_name = domain_name

    @staticmethod
    def getDomain(domain):
        domain_parts = domain.split('.')
        if len(domain_parts) > 2:
            dirpath = os.path.dirname(os.path.dirname(os.path.realpath(__file__)))
            domainfile = dirpath + "/domain.ini"
            domainarr = []
            with open(domainfile) as f:
                for line in f:
                    val = line.strip()
                    domainarr.append(val)

            rootdomain = '.'.join(domain_parts[-(2 if domain_parts[-1] in domainarr else 3): ])
            selfdomain = domain.split(rootdomain)[0]
            return (selfdomain[0:len(selfdomain)-1], rootdomain)
        return ("", domain)

    def curl(self, url, data, method):
        if sys.version_info[0] < 3:
            print ("s")
            import urllib2
            from urllib2 import URLError, HTTPError
            httpdata = json.dumps(data).encode('utf-8')
            req = urllib2.Request(url=url, data=httpdata)
            req.get_method = lambda: method
            req.add_header('accept', 'application/json')
            req.add_header('Content-Type', 'application/json')
            key = "sso-key " + self.access_key_id + ':' + self.access_key_secret
            req.add_header('authorization', key)
            try:
                with urllib2.urlopen(req) as res:
                    code = res.getcode()
                    print (res.info())
                    resinfo = res.read().decode('utf-8')
                    result = True
                    if code != 200:
                        result = False
                    return (result, resinfo)
            except AttributeError as e:
                #python2 处理 PATCH HTTP 方法的一个Bug，不影响结果
                return (True,'')
            except (HTTPError, URLError) as e:
                return (False, str(e))

        else :
            import urllib.request
            from urllib.error import URLError, HTTPError
            
            httpdata = json.dumps(data).encode('utf-8')

            req = urllib.request.Request(url=url, data=httpdata, method=method)
            req.add_header('accept', 'application/json')
            req.add_header('Content-Type', 'application/json')
            key = "sso-key " + self.access_key_id + ':' + self.access_key_secret

            req.add_header('authorization', key)
            try:
                with urllib.request.urlopen(req) as res:
                    code = res.getcode()
                    # print (res.info())
                    resinfo = res.read().decode('utf-8')
                    result = True
                    if code != 200:
                        result = False
                    return (result, resinfo)
            except (HTTPError, URLError) as e:
                return (False, str(e))

    def CreateDNSRecord(self, name, value, recordType='TXT'):
        url = "https://api.godaddy.com/v1/domains/" + \
            self.domain_name + "/records"
        data = [{"data": value, "name": name, "ttl": 3600, "type": recordType}]
        return self.curl(url, data, "PATCH")

    def GetDNSRecord(self, name, recordType='TXT'):
        url = "https://api.godaddy.com/v1/domains/" + \
            self.domain_name + "/records/" + recordType + "/" + name
        return self.curl(url, {}, "GET")

    def DeleteDNSRecord(self, name, recordType='TXT'):
        '''
        Godaddy DNS  没有提供删除DSN记录的API
        '''
        return True

file_name, cmd, certbot_domain, acme_challenge, certbot_validation, ACCESS_KEY_ID, ACCESS_KEY_SECRET = sys.argv

certbot_domain = GodaddyDns.getDomain(certbot_domain)
if certbot_domain[0] == "":
    selfdomain = acme_challenge
else:
    selfdomain = acme_challenge + "." + certbot_domain[0]

domain = GodaddyDns(ACCESS_KEY_ID, ACCESS_KEY_SECRET, certbot_domain[1])
# print (domain.GetDNSRecord(selfdomain))

if cmd == "add":
    print(domain.CreateDNSRecord(selfdomain, certbot_validation))
